---
title: "PD-1 Responsive Glioblastoma Case Study: Differential Gene Expression & Pathway Analysis"
author: "Paula Restrepo"
date: "May 2019"
output: 
  html_document: 
    df_print: tibble
    highlight: kate
    number_sections: yes
    theme: spacelab
    toc: yes
    toc_float: true
---


Note: Some of the figures in this document use grid.arrange or ggarrange. The heights/widths arguments in grid.arrange() and ggarrange() may have to be tweaked to get the proportions of the panels to look less lopsided.

# Setup
```{r include=FALSE}

libs <- c('DESeq2', 'edgeR', 'limma', 'tidyverse', 'combinat', 'ggrepel','gridExtra', 'grid',
          'org.Hs.eg.db', 'AnnotationDbi', 'ggpubr', 'fgsea', 'ggbeeswarm', # 'MESS', 'pbmcapply', # (for running TCGA bootstrap resampling)
           'lattice')
lapply(libs, library, character.only=TRUE)
rm(libs)

# set up themes for ggplotting
theme <- theme(legend.position = 'bottom',
               legend.justification = 'left',
               text = element_text(size=16),
               axis.text = element_text(size=16, hjust = 1, face='bold'),
               axis.title = element_text(size=18, face = "bold"),
               legend.title= element_text(size=18, face='bold'),
               strip.text=element_text(size=18, face='bold'),
               plot.title=element_text(size=18, face='bold'))
# load inputs

load("../data/001_DEGsAndPathwayAnalysis_inputs.rdata")
load("../data/tcga_exprs_integration.rdata")
ind <- read_tsv('../data/tcga_exprs_integration.tsv.gz')
  
  
cbpal <-  c("#CC79A7", # purple, primary
            "#E69F00", # yellow, A
            "#56B4E9", # blue, B
            "#009E73", # green, C
            "#7f7f7f") #gray, tcga

source("functions.r")
```


# DEG analysis

## Create design matrix


Create  design matrix using all permunations of sample codes. 1 and 0 corresponds to which side of the comparison the sample in the sample_names column is. The sample names will just correspond to the rownames of the phenotype table. this gets combined in a for-loop later.

This design matrix is referred to as the "pooled comparison strategy" used in the analysis, as it allows us to look at the differences between different regions without resorting entirely to 1v1 comparisons. This way, we can identify gene signatures unique to a particular region compared to the rest, or to identify gene signatures that may be associated with samples containing similar clonal architectures from the previous clonevol analysis.

```{r }
d1 <- ((combinat::permn(as.character(c('1', '1', '0', '0')))))
d1 <- t( unique( data.frame( matrix(unlist(d1), nrow = factorial(4), byrow = T))))
class(d1) <- "numeric"

d2 <- ((combinat::permn(as.character(c('1', '1', '1', '0')))))
d2 <- t( unique( data.frame( matrix(unlist(d2), nrow = factorial(4), byrow = T))))
class(d2) <- "numeric"

design <- cbind(d1, d2)
colnames(design) <- NULL
rownames(design) <- NULL

# clean up intermediaries
rm(d1, d2)
```


## Generate DEG results using pooled comparisons in DESeq2

```{r message=FALSE, warning=FALSE}
comparison_list <- NULL # store the comparisons here.
deg_list <- NULL # store the results of the comparison here.
# loop through the combinations in the design matrix to do permuted DEG tests in pooled
# comparisons using DESeq2.
for (i in 1:ncol(design)){

    sample_code <- design[, i]
    ph <- data.frame(sample_name=rownames(case.phenotype), sample_code=sample_code)

    dds <- DESeqDataSetFromMatrix(countData = gcounts, colData = ph, design = ~sample_code)
    dds <- DESeq(dds)
    res <- results(dds)
    resOrdered <- as.data.frame(res[order(res$padj),])
    resOrdered$ensid <- rownames(resOrdered)
    deg_list[[i]] <- resOrdered
    comparison_list[[i]] <- ph
}

# clean up intermediaries
rm(resOrdered, dds, res, ph, sample_code)
```

## Clean up sample labels for visualization

```{r }
# now take the comparison list into something easier to coerce into a sample v sample string
# for the volcano plot labels
comparison_names <- NULL

for (i in 1:length(comparison_list)){
    comparison <- comparison_list[[i]]
    s0 <- paste(as.character(comparison[comparison$sample_code == 0,]$sample_name), collapse = ', ')
    s1 <- paste(as.character(comparison[comparison$sample_code == 1,]$sample_name), collapse=', ')

    deg_list[[i]]$sample_comparison <- as.character(paste(s0, s1, sep=' - vs - '))
    
    comparison_names[i] <- as.character(paste(s0, s1, sep=' - vs - '))
}

names(deg_list) <- comparison_names
rm(s0, s1)

# get top 25 p values and top/bottom logFC values
top_genes <- lapply(deg_list, function(x){
    x <- x %>%  filter(!is.na(padj)) %>%
        filter(padj < 0.5) %>%
        left_join(genes, by=c('ensid'))
    x1 <- x %>% top_n(15, wt=as.numeric(-log(padj)))
    x2 <- x %>% filter(log2FoldChange > 1) %>% top_n(10, wt=abs(log2FoldChange))
    x3 <- x %>% filter(log2FoldChange < -1) %>% top_n(10, wt=abs(log2FoldChange))
    x <- unique(rbind(x1, x2, x3))

    rm(x1, x2, x3)
    return(x)
})

top_genes <- do.call(rbind, top_genes)
```


## Remove mirrored comparisons


```{r fig.height=10, fig.width=15}
# coerce results to data frame and remove mirrored comparisons
mirrored_comparisons <- c("C, P - vs - A, B", "B, C - vs - A, P", "B, P - vs - A, C")
deg_df <- do.call(rbind, deg_list)
deg_df <- deg_df %>% as_tibble() %>%
    filter(!is.na(padj)) %>%
    filter(!sample_comparison %in% mirrored_comparisons) %>%
    left_join(genes, by=c('ensid'))

top_genes <- top_genes %>%
  filter(!is.na(padj)) %>%
  filter(!sample_comparison %in% mirrored_comparisons)

summary(as.factor(top_genes$sample_comparison))
```

## Visualize with volcano plot

```{r fig.height=8, fig.width=25}
volcano_plot_filt <- ggplot(data=deg_df[! deg_df$sample_comparison %in% c('A - vs - B, C, P', 'A, P - vs - B, C'),]) + aes(x=log2FoldChange, y=-log(padj), color=biotype) +
    geom_point() +
    geom_label_repel(data=top_genes[! top_genes$sample_comparison %in% c('A - vs - B, C, P', 'A, P - vs - B, C'),], aes(label=gene), show.legend=FALSE,  force=3, size=4) +
    facet_wrap(~ sample_comparison, scales='free',nrow=1) +
    theme_classic() + theme

volcano_plot_filt
```


## Get GSEA scores for MSigDB oncogenic signatures

First coerce the deg results from DEseq2 into a list of ranks for each comparison using the method described here: https://stephenturner.github.io/deseq-to-fgsea/

Then, we will use fgsea to perform gene set enrichment on all the MSigDB oncogenic signatures.


```{r }
deg_ranks <- lapply(deg_list, function(x){
    x %>%
        left_join(genes, by=c('ensid')) %>%
        na.omit() %>%
        distinct() %>%
        group_by(gene) %>%
        summarise(stat=mean(stat)) %>%
        deframe()
    } )

deg_ranks <- deg_ranks[! names(deg_ranks) %in% mirrored_comparisons]
deg_ranks <- deg_ranks[! names(deg_ranks) %in% c("A - vs - B, C, P","A, P - vs - B, C")]

load('../data/msigdb_gene_signatures.rdata')
```

Generate the enrichment scores & coerce to dataframe

```{r }
# compute gene set enrichments
all_fgsea <- list(hallmark_fgsea = gsea_lister(rank_list = deg_ranks, sig_list = hallmark_gsigs, sig_name='HALLMARK'),
                  kegg_fgsea = gsea_lister(rank_list = deg_ranks, sig_list = kegg_gsigs, sig_name = 'KEGG'),
                  reactome_fgsea = gsea_lister(rank_list = deg_ranks, sig_list = reactome_gsigs, sig_name = 'REACTOME'),
                  biocarta_fgsea = gsea_lister(rank_list = deg_ranks, sig_list = biocarta_gsigs, sig_name = 'BIOCARTA')
)

rm(hallmark_gsigs, kegg_gsigs, reactome_gsigs, biocarta_gsigs)
```

Select the top pathways, filter irrelevant terms, and order the pathways via hierarchical clustering.
```{r }
# select top pathways & order pathways via hclust
data_list <- lapply(all_fgsea, function(x){
  sig_name <- levels(as.factor(x$signature_name))
  x$pathway <- gsub(paste0(sig_name, '_'), '', x=x$pathway)
  x$sample_comparison <- factor(x$sample_comparison, levels=names(deg_ranks))
  
  df <- x %>%
  filter(padj < 0.05) %>% arrange(padj) %>% group_by(sample_comparison) %>%
  top_n(10, wt=-padj) %>% top_n(10, wt=abs(NES)) %>%
  arrange(abs(NES), -padj) 

  tmp <- df %>% dplyr::select(sample_comparison, pathway, NES) %>%
  group_by(sample_comparison) %>% tidyr::spread(pathway, NES) %>% 
  as.data.frame()
  
  names <- tmp$sample_comparison
  tmp$sample_comparison <- NULL
  tmp <- as.data.frame(lapply(tmp, function(x){
    x[is.na(x)] <- 0
    return(x) }))
  
  rownames(tmp) <- names
  order <- hclust( dist(t(tmp), method = "euclidean"), method = "ward.D" )$order
  df$pathway <- factor(df$pathway,levels=colnames(tmp)[order])


  df <- df[!is.na(df$pathway),]
  df <- df[df$pathway != 'NA',]

  rm(names, tmp)

  return(df)
})

# define terms that are not relevant, and will be hidden in the resulting visualization
terms <- c('INFLUENZA', 'CELL_CYCLE', 'MITOTIC', 'S_PHASE', 'G1_S_TRANSITION', 'ASTHMA', 'AUTOIMMUNE_THYROID_DISEASE',
'INFECTION','DISEASE', 'INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION','METABOLISM', 'APICAL_SURFACE','GENESIS',
 '43S', 'SYNAP','NEUROTRANSMITTER', 'GABA', 'GLUTAMATE', 'POTASSIUM_CHANNELS', 'CHANNEL_TRANSPORT', 'CELL_LINEAGE',
 'LUPUS', 'NEURONAL',  'RP_DEPENDENT_COTRANSLATIONAL_PROTEIN_TARGETING_TO_MEMBRANE', 'TRANSLATION', 'proteasome', 'complement', 'homologous_recombination',
"peptide_chain",'second_messenger','NONSENSE_MEDIATED_DECAY', 'G2M_checkpoint', 'G2_M_CHECKPOINT', 'IMMUNOREGULATORY_INTERACTIONS_BETWEEN',
'RIBOSOME', 'TIGHT_JUNCTION') 

# apply filters
data_filt <- lapply(data_list, function(x){

  y <- x
  for (term in terms){
    y <- y %>% filter(!grepl(term, x=pathway, ignore.case=TRUE))
  }

df <- y

tmp <- df %>% dplyr::select(sample_comparison, pathway, NES) %>%
  group_by(sample_comparison) %>% tidyr::spread(pathway, NES) %>% 
  as.data.frame()
  
  names <- tmp$sample_comparison
  tmp$sample_comparison <- NULL
  tmp <- as.data.frame(lapply(tmp, function(x){
    x[is.na(x)] <- 0
    return(x) }))
  
  rownames(tmp) <- names
  order <- hclust( dist(t(tmp), method = "euclidean"), method = "ward.D" )$order
  df$pathway <- factor(df$pathway,levels=colnames(tmp)[order])


  df <- df[!is.na(df$pathway),]
  df <- df[df$pathway != 'NA',]

  rm(names, tmp)

  return(df)

})
```

## Visualize gene set enrichments as a heatmap

Make one plot per gene set database and visualize as a set of stacked heatmaps.
```{r fig.height=20, fig.width=16, message=FALSE, warning=FALSE}
# set up themes for ggplotting
theme <- theme(text = element_text(size=16),
               axis.text.y = element_text(size=18, face='bold'),
               axis.text.x = element_text(size=18, face='bold'),
               axis.title = element_text(size=24, face = "bold"),
               legend.text= element_text(size=18),
               legend.title= element_text(size=24, face='bold'),
               strip.text=element_text(size=24, face='bold'),
               plot.title=element_text(size=28, face='bold'))

plot_list <- lapply(data_filt, function(x){
  plot <- ggplot(data=x) +
  aes(x=sample_comparison, y=pathway, fill=as.numeric(NES)) +
  geom_tile(color='white', size=1) +
  theme_classic() +
  theme + theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  scale_x_discrete(drop=FALSE) +
  labs(x ='', y=levels(as.factor(x$signature_name))) +
  scale_fill_gradient2(low = "lightseagreen", mid='white', high = "tomato", name="NES")  +
  guides(fill = guide_colourbar(barwidth = 1.5, barheight = 25)) 
  return(plot)
})

plot_list[[4]] <- plot_list[[4]] + theme_classic() + theme + rotate_x_text(angle=45)

names(plot_list) <- c('HALLMARK', 'KEGG', 'REACTOME', 'BIOCARTA')
pathway_hm <- ggarrange(plotlist=plot_list, nrow=length(all_fgsea), align='v',  common.legend=TRUE, legend='right', heights=c(1.2,0.6,0.5,1.6))

pathway_hm
```


# Integrate TCGA expression information

## PCA

```{r}
case_int_pcs <- ind
```

```{r}
tcga_pca <- ggplot(data = case_int_pcs) + 
  aes(x=Dim.1, y=Dim.2, color=sample, label=label) +
  geom_density2d(color="#7f7f7f", bins=50, alpha=0.7) + 
  geom_point(alpha=0.7, size=5) +
  geom_label_repel(size=8, show.legend=FALSE, force=10, min.segment.length = 10, segment.size = 1) +
  labs(x='PC-1: 14.3% of variance explained', y='PC-2: 9% of variance explained') +
  theme_classic() + theme(plot.margin=margin(0.5,0.5,0.5,0.5, "in")) +
  theme + scale_color_manual(values = cbpal)
```

```{r  fig.height=7, fig.width=7, warning=FALSE}
tcga_pca
```

## With distributions of specific checkpoint and immune genes

Objective: get a measure (a p-value) of whether the variance between my four samples is greater than if you were to take the variance four random TCGA samples. To do this, do resamplings of 10% of the total possible combinations of 4 tcga-GBM samples from a pool of 155 samples.
```{r}
checkpoint.genes <- c("PD1","PDL1","PDL2","TIM3","CD28","CD3","CD4","FOXP3","DCLAMP")
val_cols <- checkpoint.genes
chk <- ind
chk$submitter_id <- as.character(chk$submitter_id)
```

```{r eval=FALSE}
# separate out samples from case study
ref_df <- chk[1:4,]
ref_df <- data.frame(var=sapply(ref_df[, val_cols], var),
                      samp_type=as.character('test'),
                      val=val_cols)

set.seed(0)

length(combn(chk[- c(1:4),]$submitter_id, 4, simplify=FALSE))
boot_pool <- chk[- c(1:4),]
# grab all unique combinations of 4 from the pool of 155 TCGA samples
# and then only take 10% of that cause there's too many
combinations <- combn(boot_pool$submitter_id, 4, simplify=FALSE)
combinations <- base::sample(combinations, length(combinations)*0.1)
```


```{r eval=FALSE}
# function to take variance of each combination of 4 for each gene/column
boot_fx <- function(x){
  # grab the combination of 4
  x <- boot_pool %>%
       filter(submitter_id %in% x) %>%
       dplyr::select(val_cols)
  # take the variance for each gene 
  x <- data.frame(var=sapply(x, var),
                  val=val_cols,
                  samp_type=as.character('random_boot'))
  return(x)
}

# do the resampling; using 3 cores (but the more the merrier!), and then bind the rows.
boot_pool <- bind_rows(pbmclapply(combinations, boot_fx, mc.cores=2))
all_boots <- bind_rows(ref_df, boot_pool)
```

Get a measure of whether the variance between my four samples is greater than if you were to take four random TCGA samples. 

```{r eval=FALSE}
pvals <- lapply(levels(factor(all_boots$val)), function(i){
    x <- all_boots %>% filter(val == i) %>% as.data.frame()
    test_point <- x %>% filter(samp_type == "test") %>% dplyr::select(var) %>% deframe()
    z <- stats::density(x$var, n=length(x$var))
    xval <- as.numeric(as.matrix(as.numeric(z$x)))
    yval <- as.numeric(as.matrix(as.numeric(z$y)))

    # test right-sided
    rpval <- auc(x=xval, y =yval, from=test_point, to=max(xval) )
    # test left-sided
    lpval <- auc(x=xval, y =yval, from=min(xval), to=test_point )
    # pick the smallest of the two left-sided and right-sided tests
    # to get the auc (p-value) for the closest end of the distribution
    result <- min(c(rpval, lpval))
    return(result)
})

names(pvals) <- levels(factor(all_boots$val))
pvals <- unlist(pvals)

pvaldf <- data.frame(pval=pvals,
                    gene=names(pvals))
```


## plot the checkpoint violins
```{r eval=FALSE}
ind2 <- tidyr::gather(chk, gene, cpm, val_cols)
ind2$gene <- as.factor(ind2$gene)
ind2$cpm <- as.numeric(ind2$cpm)


ind2 <- ind2 %>% inner_join(pvaldf, by=c('gene'))
ind2$sample <- as.factor(ind2$sample)
```
Note: we saved a pre-computed version of this to reduce runtime and will use it in this RMarkdown document.
```{r}
ind2 <- read_tsv('../data/bootstrap_tested_tcga_checkpoint_genes_long.tsv.gz')
```

```{r}
checkpoint_violins <- ggplot(data=ind2) + aes(y=cpm, x=gene, color=sample) + 
  #geom_violin(alpha=0.05, aes(color=NULL), fill="#7f7f7f", show.legend = FALSE) + 
  geom_boxplot(width=0.5, aes(color=NULL, fill=NULL), show.legend = FALSE ) +
  geom_beeswarm(data=ind2[ind2$sample =='TCGA-GBM',], alpha=0.4, size=2, show.legend=FALSE) +
  geom_point(data=ind2[ind2$sample !='TCGA-GBM',], size=6, alpha=0.75) +
  theme_classic() + theme + labs(y='Voom Normalized Expression', x='', color='Sample') + 
  # don't add duplicate labels
  geom_text(data=ind2[(ind2$pval<0.05) & (ind2$sample=='Primary'),],
            y=-5,size=6, show.legend=FALSE,
            aes(x=gene, label=paste('P =',round(pval, digits=4)), color=NULL)) +
            scale_color_manual(values = cbpal, aesthetics = 'color')
```

```{r  fig.height=8, fig.width=20, message=FALSE, warning=FALSE}
checkpoint_violins
```

# MORE PCAs

## color based on survival

```{r}
surv_pca <- ggplot(data = case_int_pcs) + 
  aes(x=Dim.1, y=Dim.2, color=MonthsOverallSurvival, label=label, shape=VitalStatusFactor) +
  geom_density2d(color="#7f7f7f", bins=50, alpha=0.7, aes(shape=NULL, color=NULL)) + 
  geom_point(alpha=0.7, size=5) +
  geom_label_repel(size=8, show.legend=FALSE, force=10, min.segment.length = 10, segment.size = 1) +
  labs(x='PC-1: 14.3% of variance explained', y='PC-2: 9% of variance explained') +
  theme_classic() + 
  theme(plot.margin=margin(0.5,0.5,0.5,0.5, "in")) +
  theme + scale_color_gradient(low="white", high="black")
```

```{r  fig.height=7, fig.width=7, warning=FALSE}
surv_pca
```

## color based on checkpoint gene

```{r}
gene_pca <- ggplot(data = ind2) + 
  aes(x=Dim.1, y=Dim.2, color=cpm, label=label) +
  geom_density2d(color="#7f7f7f", bins=50, alpha=0.7) + 
  geom_point(alpha=0.7, size=5) +
  geom_label_repel(size=8, show.legend=FALSE, force=10, min.segment.length = 10, segment.size = 1) +
  labs(x='PC-1: 14.3% of variance explained', y='PC-2: 9% of variance explained') +
  facet_wrap(~ gene) +
  theme_classic() + 
  theme(plot.margin=margin(0.5,0.5,0.5,0.5, "in")) +
  theme + scale_color_gradient(low="white", high="blue", name='Normalized Expression')
```

```{r  fig.height=15, fig.width=15, warning=FALSE}
gene_pca
```
# Manuscript Figure 4

```{r, fig.height=18, fig.width=34}
tcga_integration <- ggarrange(tcga_pca, checkpoint_violins, nrow=2, heights=c(2,0.7), align='v', common.legend=TRUE, legend='right')

pdf('fig4_withCheckpointViolins.pdf', height=18, width=34)
grid.arrange(pathway_hm, tcga_integration, ncol=2, widths=c(0.8, 1.25))
dev.off()
```

```{r}
sfig3 <- ggarrange(surv_pca, gene_pca, ncol=2)
```


# RSession Info
```{r}
sessionInfo()
```

