---
title: "PD-1 Responsive Glioblastoma Case Study: Mutation & CNV analysis"
author: "Paula Restrepo"
date: "May 2019"
output: 
  html_document: 
    df_print: tibble
    highlight: kate
    number_sections: yes
    theme: spacelab
    toc: yes
    toc_float: true
---

# Set up
```{r include=FALSE}
libs <- c('tidyverse', 'ggrepel', 'org.Hs.eg.db', 'AnnotationDbi', 'RColorBrewer', 'ggpubr', 'GenomicRanges', 'sciClone', 'clonevol', 'reshape2')
lapply(libs, library, character.only=TRUE)
rm(libs)
```


```{r message=FALSE, warning=FALSE}
set.seed(0)

# set up themes for ggplotting
theme <- theme(text = element_text(size=24),
               axis.text.y = element_text(size=24, hjust = 1, face='bold'),
               axis.text.x = element_text(size=24, hjust = 1,angle=45,  face='bold'),
               axis.title = element_text(size=28, face = "bold"),
               legend.text= element_text(size=24),
               legend.title= element_text(size=28, face='bold'),
               strip.text=element_text(size=28, face='bold'),
               legend.position='right',
               legend.justification='left')

# load inputs
somatic_mutations <- read_tsv("../data/somatic_mutations_table.tsv.gz")
cns_table <- read_tsv("../data/cns_table.tsv.gz")
load("../data/cosmic.cgc.rdata")

source('functions.r')
```

# Create an Oncoprint chart using ggplot
## set up mutation chart

Select genes to visualize in heatmap.
```{r}
# manually selected set of genes to look for
gene_list1 <- c("MAPK1", 'MAPK2','MAPK4', "MAP3K1", "MAPKAPK5", "MAP7", "MAP3K7",
               'PTEN', 'EGFR', 'TP53','BRAF', 'PTPN11',"PTPN3","PTPN4","PTPN5", 'PTPRH',
               'MTOR', 'RHOA', 'AKT1', "AKT2", "AKT3", "PIK3CA", "PIK3CB","IDH1",
               "USH2A", "MLL3", "BRCA1", "POLE1", "NF1", "ATRX", "CIC", "CDK11A" )

top_genes <- somatic_mutations %>% 
  arrange(desc(as.numeric(gt_AF))) %>%
  group_by(sample) %>% top_n(n=25, wt=gt_AF)

gene_list <- c(gene_list1, 
               top_genes$Gene_refGene,
               levels(factor(cosmic.v89.cgc$`Gene Symbol`)))

mut_heatmap_df <- somatic_mutations %>%
  filter(Gene_refGene %in% gene_list1) %>%
  mutate(sample=fct_relevel(sample, c('Primary', "Recurrent_A",
                                      'Recurrent_B', 'Recurrent_C')),
         gene=Gene_refGene)
```

Prepare tumor mutation burden df for visualization
We use the formula of TMB = Number of mutations / Megabases of DNA covered by the sequencing assay; in the case of the Agilent SureSelect library it's about 54 MB.
```{r}
tmb_df <- somatic_mutations %>%
  group_by(sample, translational_effect) %>%
  summarise(count = n(),tmb=count/54) %>%
  ungroup() %>%
  mutate(sample = fct_relevel(sample, c('Primary', "Recurrent_A", 
                                        'Recurrent_B', 'Recurrent_C')) )

tmb2 <- somatic_mutations %>% group_by(sample) %>%
  summarise(count=n(),tmb=count/54) %>%
  ungroup() %>%
  mutate(sample =  fct_relevel(sample, c('Primary', "Recurrent_A",
                                         'Recurrent_B', 'Recurrent_C')) )

tmb_df
tmb2
```

## set up cnv chart

Select genes covered by CNVs that we wish to visualize. Define bins of CNV severity. We use a +/- 0.25 log2 threshold under which we would consider a call to be neutral or not cnv-altered.

```{r }
cnv_heatmap_df <- cns_table %>%
  filter(abs(log2) >= 0.25)

cnv_heatmap_df <- lapply(gene_list, function(x){
  y <- cnv_heatmap_df[grepl(x, cnv_heatmap_df$gene, ignore.case = TRUE), ]
  if (nrow(y)!= 0) {
    y$query_gene <- x
  }
  return(y)
})

cnv_heatmap_df <- do.call(rbind, cnv_heatmap_df)
cnv_heatmap_df <- unique(cnv_heatmap_df)
cnv_heatmap_df$query_gene <- as.character(cnv_heatmap_df$query_gene)

cnv_heatmap_df <- cnv_heatmap_df %>% 
    group_by(query_gene, sample) %>% 
    summarise(average_cnv=mean(log2)) %>%
    ungroup()

cnv_heatmap_df$sample <- as.factor(cnv_heatmap_df$sample)
levels(cnv_heatmap_df$sample) <- c('Recurrent_A', 'Recurrent_B', 'Recurrent_C', 'Primary')

cnv_list <- c(gene_list1, 'SND1', "MIER2", "VAV1")
cnv_heatmap_df <- cnv_heatmap_df %>% filter(query_gene %in% cnv_list)
cnv_heatmap_df$cnv_type <- cut(cnv_heatmap_df$average_cnv, include.lowest = TRUE, breaks = c(-10,-1, -0.25, 0.25, 1, 10))
levels(cnv_heatmap_df$cnv_type) <- c('deletion', 'mild deletion', 'neutral', 'mild amplification', 'amplification')

cnv_heatmap_df$sample <- fct_relevel(cnv_heatmap_df$sample, c('Primary', "Recurrent_A", 'Recurrent_B', 'Recurrent_C'))

cnv_heatmap_df <- cnv_heatmap_df %>% filter(cnv_type != 'neutral')

cnv_heatmap_df
```

Calculate the genomic instability index (% genome altered) for each sample using the 54 Mb coverage and 0.5-fold change threshold for neutrality
```{r }
case_cnv_granges <- lapply(levels(factor(cns_table$sample)), function(x){
  x <- cns_table %>% filter(sample %in% x)
  
  x <- makeGRangesFromDataFrame(x[abs(x$log2) > 0.5 ,], keep.extra.columns = TRUE,
                                seqnames.field = 'chromosome', start.field = 'start', end.field = 'end')
  
  return(x)
  
  })


case_cnv_granges <- GRangesList(case_cnv_granges)
case_total_cnv_lengths <- sum(width(reduce(case_cnv_granges, ignore.strand=TRUE)))
cnv_lenths_mb <- case_total_cnv_lengths / (1e7)
case_genomic_instability <- as.data.frame(abs(case_total_cnv_lengths))
case_genomic_instability$cnv_lenths_mb <- cnv_lenths_mb
case_genomic_instability$instability_idx <- (cnv_lenths_mb / 54) * 100
rm(case_total_cnv_lengths, case_cnv_granges, cnv_lenths_mb)

pheno_summary <- case_genomic_instability

# as calculated from TheTA2 (presented as a proportion)
pheno_summary$tumor_purity <- c(36.6, # A
    37.6, #B
    30.2, #C
    38.2 ) #PRIMARY

pheno_summary$sample <- as.factor(c("Recurrent_A", 'Recurrent_B', 'Recurrent_C', 'Primary'))
pheno_summary$sample <- fct_relevel(pheno_summary$sample, c('Primary', "Recurrent_A", 'Recurrent_B', 'Recurrent_C'))

pheno_summary <- pheno_summary %>% tidyr::gather(var, val, c('tumor_purity', 'instability_idx'))
pheno_summary$var <- as.factor(pheno_summary$var)
levels(pheno_summary$var) <- c('Genomic Instability Index', 'Tumor Purity')
pheno_summary$val <- round(pheno_summary$val, digits=1)

```

## plot the mutation summary
```{r fig.height=20, fig.width=15, message=FALSE, warning=FALSE}
mut_burden_barplot <- ggplot(data=tmb_df) + 
  aes(x=sample,y=tmb, fill=translational_effect) + 
  geom_bar(stat='identity') +
  geom_text(data=tmb2, aes(label=paste('N =',count), fill=NULL),size=8.5, nudge_y = 0.5, show.legend=FALSE) +
  labs(x='', y='TMB (Muts/MB)', fill='Translational Effect') +
  theme_classic() +
  theme +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
   scale_fill_manual(values = c('grey30', "grey70"))

mut_heatmap <- ggplot(data=mut_heatmap_df) +
  aes(x=sample, y=gene, fill=variant_class) +
  geom_tile(color='white', size=1, alpha=0.9) +
  labs(x='',y='SNV', fill='SNV Type') +
  theme_classic() +
  theme +
  scale_fill_brewer(palette="Spectral") +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

cnv_heatmap <- ggplot(data=cnv_heatmap_df) +
  aes(x=sample, y=query_gene, fill=cnv_type) +
  geom_tile(color='white', size=1) +
  labs(x='',y='CNV', fill='CNV Type') +
  theme_classic() +
  theme +
  scale_fill_manual(values = rev(brewer.pal(5,"RdBu")),drop=FALSE, labels = c("deletion", "", "neutral", "", "amplification"))

phenotype <- ggplot(data=pheno_summary) +
  aes(x=sample, y=var, fill=val) +
  geom_tile(color='white', size=1) +
  geom_text(aes(label=paste0(val, '%')), show.legend=FALSE, size=8.5) +
  labs(x='',y='', fill = "Percent") +
  theme_classic() +
  theme +theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())+
  scale_fill_gradientn(colors = brewer.pal(100,"PuBu"), limits = c(0,60), breaks = c(0,30,60),) 


oncoprint_chart <- ggarrange(mut_burden_barplot ,phenotype,mut_heatmap,cnv_heatmap, nrow=4, align = 'v', heights = c(1.2,1,4,5))

oncoprint_chart
```

## Find the overlap of mutations

Calculate the overlap
```{r message=FALSE, warning=FALSE}
count <- function(x,y){ f<- table(rbind(x,y)); sum(f[f>1])/sum(f) }

hm.palette <- colorRampPalette(rev(brewer.pal(100, 'YlGnBu')), space='Lab')
hm.theme <-  theme(legend.position='bottom',
                legend.justification='left',
                panel.background = element_rect(fill="white", color='white'),
                axis.text.x = element_text(angle = 45, hjust = 1, size=25, face="bold", color='black'),
                axis.text.y = element_text(hjust = 1, size=24, face="bold", color='black'),
                text = element_text(size=25))


vcf_filtered_df_OL <- somatic_mutations
mutation_overlap <- NULL

for (sample in levels(as.factor(vcf_filtered_df_OL$sample))){
    idx <- vcf_filtered_df_OL$sample == sample
    mutation_overlap[[sample]] <- unique(as.matrix(factor(vcf_filtered_df_OL[idx,]$variant)))
}

mutation_overlap <- sapply(mutation_overlap, function(x) sapply(mutation_overlap, function(y) count(x,y)))
rownames(mutation_overlap) <- levels(as.factor(vcf_filtered_df_OL$sample))
colnames(mutation_overlap) <- levels(as.factor(vcf_filtered_df_OL$sample))
mutation_overlap2 <- melt( mutation_overlap)
colnames(mutation_overlap2) <- c('sample1', 'sample2', 'overlap')

summary(mutation_overlap2)
```

plot the overlap

```{r fig.height=10, fig.width=10, message=FALSE, warning=FALSE}
# moverlap plot

moverlap <- ggplot(data=mutation_overlap2) + aes(x = sample1, y = sample2) +
            geom_tile(color='black', size=0.5, aes(fill = round(overlap * 100, 2))) +
            geom_label(aes(label = round(overlap * 100, 2)), size=9) +
            labs(x='', y='') +
            scale_fill_gradientn(colours = hm.palette(100), name='% Overlap', limits= c(0,100)) +
            theme_minimal() + hm.theme +
            guides(fill = guide_colorbar(barwidth = 20))

moverlap
```


# RSession Info
```{r}
sessionInfo()
```
